pkg/algorithm/hived_algorithm.go

```diff
@@ -55,15 +55,19 @@ type HivedAlgorithm struct {
 	vcFreeCellNum map[api.VirtualClusterName]map[CellChain]map[CellLevel]int32
 	// total number of preassigned free cells in all the VCs of each cell type
 	allVCFreeCellNum map[CellChain]map[CellLevel]int32
-	// number of cells left in the physical cluster of each cell type
-	// (i.e., the cell is either in the free list or could be obtained by splitting such a free cell.)
+	// Number of cells left in the physical cluster of each cell type
+	// (i.e., the cell is either in the free list or could be obtained by splitting such a free cell).
+	// For example, we have a level-2 cell in the free list (which can be split into 2 level-1 cells).
+	// Then the left cell numbers will be level-2: 1, level-1: 2.
 	totalLeftCellNum map[CellChain]map[CellLevel]int32
-
-	// number of bad free cells in the physical cluster of each cell type
-	badFreeCellNum map[CellChain]map[CellLevel]int32
-	// number of free cells in each VC that are doomed to be bad (i.e., when the healthy
+	// bad free cells in the physical cluster of each cell type
+	badFreeCells map[CellChain]ChainCellList
+	// free cells in each VC that are doomed to be bad (i.e., when the healthy
 	// free cells in the physical cluster is fewer that the VC's free cells)
-	vcDoomedBadCellNum map[api.VirtualClusterName]map[CellChain]map[CellLevel]int32
+	vcDoomedBadCells map[api.VirtualClusterName]map[CellChain]ChainCellList
```

Comment:
Mentioned that how we define bad cell.
And the doomed bad cells may contain healthy child cell

Reply:
"A bad cell may contain healthy children" is a common property of all bad cells, not just for doomed bad cell. so in the comment I do not emphasize doomed bad cell for this

Reply:
Mentioned vc may also have doomed healthy cell

---
In reply to: [437884123](https://github.com/microsoft/hivedscheduler/pull/17#discussion_r437884123) [](ancestors = 437884123)